zeroex/orderwatch/order_watcher.go

```diff
@@ -940,8 +865,55 @@ func (w *Watcher) add(orderInfos []*ordervalidator.AcceptedOrderInfo, validation
 		}
 	}
 
-	// TODO(albrow): How to handle the edge case of orders that were not
-	// added due to the max expiration time changing?
+	// HACK(albrow): We need to handle orders in the orderInfos argument that
+	// were never added due to the max expiration time effectively changing
+	// within the database transaction above. In other words, new orders that
+	// _were_ added can change the effective max expiration time, meaning some
+	// orders in orderInfos were actually not added. This should not happen
+	// often. For now, we respond by emitting an ADDED event immediately
+	// followed by a STOPPED_WATCHING event. If this order was submitted via
+	// RPC, the RPC client will see a response that indicates the order was
+	// successfully added, and then it will look like we immediately stopped
+	// watching it. This is not too far off from what really happened but is
+	// slightly inefficient.
+	//
+	// TODO(albrow): In the future, we should return an error and then react to
+	// that error differently depending on whether the order was received via
+	// RPC or from a peer. In the former case, we should return an RPC error
+	// response indicating that the order was not in fact added. In the latter
+	// case, we should effectively no-op, neither penalizing the peer or
+	// emitting any order events.
+	for _, orderToAdd := range orderInfos {
+		_, wasAdded := addedMap[orderToAdd.OrderHash]
+		if !wasAdded {
+			addedEvent := &zeroex.OrderEvent{
```

Comment:
In case this is still relevant with the planned changes:
We add an `addedEvent` for each order in `orderInfos`, so it'd be more clear if this loop just added the `stoppedWatchingEvent`'s.

We might not need to handle this edge case at all if we include the newly expired orders in both `removedOrders` and `addedOrders`?

Reply:
I started working on refactoring `orderwatch.Watcher.add` as we discussed in our call. But it turned out to be a huge rabbit hole and felt out of scope for this PR. Instead, I opted to keep the hack where we emit an `ADDED` event immediately followed by a `STOPPED_WATCHING` event. And I applied your first suggestion here.

I created https://github.com/0xProject/0x-mesh/issues/837 to track the tech debt associated with our order validation pipeline. It is definitely something that deserves another look.